### PR TITLE
chore: refactor cli output engine for multiple outputs.

### PR DESCRIPTION
**Pull request by akuhlens**
This PR refactors/multiplexes the `OutputHandler` in `output.py` so that it should be trivial to add multiple output destinations.

towards [SAF-341](https://linear.app/semgrep/issue/SAF-341/multiple-output-options-simultaneously)

# test plan

CI should continue to pass.

### CODE REVIEW

File: cli/src/semgrep/output.py
Comment (amchiclet) [inline cli/src/semgrep/output.py]: typo: "necessary"

For the record, I never spell this word correctly.

File: cli/src/semgrep/output.py
Comment (amchiclet) [inline cli/src/semgrep/output.py]: typo: "typechecking"

File: cli/src/semgrep/output.py
Comment (amchiclet) [inline cli/src/semgrep/output.py:152]: Personally I really wish didn't need to be backward compatible, so you can just use `outputs` and not need to have `output_format` floating around.

The need for backward compatibility seems to be a requirement from 3 years ago: https://github.com/semgrep/semgrep/commit/3737d0faf5028cf9bcfe7878797725b68db6af11

Can anyone (cc reviewers of that PR @aryx @emjin @brendongo ) give us more context about `semgrep_main.invoke_semgrep`? Can we safely break compatibility here, or no?

**Review by amchiclet — COMMENTED** [review]
I've left some initial questions about the history of output.py. I hope they get answered, and I can do another review.

File: cli/src/semgrep/commands/ci.py
Comment (amchiclet) [inline cli/src/semgrep/commands/ci.py:611]: This means that if there are multiple output formats specified, and if even one of them wants to keep ignored, then all of them need to keep ignored.

Just wanted to double-check if this is the intended behavior or not.

File: cli/src/semgrep/output.py
Comment (amchiclet) [inline cli/src/semgrep/output.py]: Optional: this is probably more about the limitations on how github diffs the changes. Once things got indented, now it looks like a big change, when I think it's just a couple of small tweaks.

I think the changes would be easier to read if you have `_build_outputs` do the for loop and yielding, and `_build_output` just return the pair (output_destination, output_format).

In general, I like it more when indented blocks aren't too long. :)

Your call.

File: cli/src/semgrep/run_scan.py
Comment (amchiclet) [inline cli/src/semgrep/run_scan.py:717]: Same comment as ci.py.

File: cli/src/semgrep/output.py
Comment (amchiclet) [inline cli/src/semgrep/output.py:233]: Actually, with the dict setup like this, is it possible that when we don't specify the output, but specify multiple formats, then will we only output to the last output format listed?

Comment (akuhlens) [reply cli/src/semgrep/output.py:233]: No, I think you would just get one of the output formats. I wasn't thinking about this case because the CLI can only specify one output right now. I was planning on requiring that each format have a distinct destination. I will put a check in the normalize step to error if duplicate destinations were given.

**Review by aryx — COMMENTED** [review]
Would be good to support this in osemgrep ..

File: cli/src/semgrep/commands/ci.py
Comment (akuhlens) [reply cli/src/semgrep/commands/ci.py:611]: This was the intended behavior but only because I couldn't see an easy way to filter things for just the output formatters. We might need to push the filtering into the output formatters later, but right now the run_scan file generates metrics off the filtered results so I didn't want to change the behavior there.

File: cli/src/semgrep/output.py
Comment (akuhlens) [reply cli/src/semgrep/output.py:152]: Even with the reply on [this thread](https://semgrepinc.slack.com/archives/C05STGMM53Q/p1713312640336109) that we can break this interface I would still like to land the code that I have working now, and clean it up in a later PR. This task is a contractual commit for the end of April which means that it needs to get into next week's release.

File: cli/src/semgrep/commands/ci.py
Comment (amchiclet) [reply cli/src/semgrep/commands/ci.py:611]: Sounds good to keep it like this now. Let's move the discussion to linear. Can you post a comment there to remind us that we should also think about `keep_ignores` for different formats?

File: cli/src/semgrep/output.py
Comment (amchiclet) [reply cli/src/semgrep/output.py:152]: I agree that we should get this PR in first.

File: cli/src/semgrep/output.py
Comment (amchiclet) [reply cli/src/semgrep/output.py:233]: Sounds good.

File: cli/src/semgrep/output.py
Comment (amchiclet) [inline cli/src/semgrep/output.py]: Heads up, my [recent PR](https://github.com/semgrep/semgrep/pull/10113) touches this code a bit, so you'll likely need to rebase. I tried to be mindful about the changes to support multiple output formats, so I think you shouldn't run into issues when rebasing, but let me know if you have questions.

**Review by amchiclet — APPROVED** [review]
Thanks. Would you mind creating a ticket to also implement this in osemgrep? I know Pad would be a bit sad it's not implemented here yet, but I understand you'd like it in time for a contractual commitment.

File: cli/src/semgrep/commands/ci.py
Comment (akuhlens) [reply cli/src/semgrep/commands/ci.py:611]: Done! ([link](https://linear.app/semgrep/issue/SAF-341/multiple-output-options-simultaneously#comment-c72f69dc))

File: cli/src/semgrep/output.py
Comment (aryx) [inline cli/src/semgrep/output.py:124]: would be nice to explain here an example of invariant, that is not enforced by OutputSettings but is by NormalizedOutputSettings.
Maybe just say like "for example, outputs is now a Dict[...] instead of Optional[Dict[...]]

**Pull request merged**
